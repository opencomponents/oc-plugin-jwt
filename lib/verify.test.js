const { keys, utils } = require('../test/utils');

const dataToTokenzie = {
  admin: true,
  name: 'John Doe',
  sub: '1234567890'
};

describe('verify', () => {
  describe('when given a valid public key signed token', () => {
    const token = utils.getTokenSignedWithPrivateKey(dataToTokenzie);

    it('returns the verified token with no errors', () => {
      const verify = require('./verify');
      verify.register({ keys }, {}, () => {});
      verify.execute(token, (error, verifiedToken) => {
        expect(error).toBe(null);
        expect(verifiedToken).toMatchSnapshot();
      });
    });
  });

  describe('when given a valid secret signed token', () => {
    const token = utils.getTokenSignedWithSecret(dataToTokenzie);

    it('returns the verified token with no errors', () => {
      const verify = require('./verify');
      verify.register({ keys }, {}, () => {});
      verify.execute(token, (error, verifiedToken) => {
        expect(error).toBe(null);
        expect(verifiedToken).toMatchSnapshot();
      });
    });
  });

  describe('when a key is not found in the keystore', () => {
    const token = utils.getTokenWithUnknownKey(dataToTokenzie);

    it('it returns an error with no token', () => {
      const verify = require('./verify');
      verify.register({ keys }, {}, () => {});
      verify.execute(token, (error, verifiedToken) => {
        expect(error).toMatchSnapshot();
        expect(verifiedToken).toBe(undefined);
      });
    });
  });

  describe('when a using a custom keystore', () => {
    const token = utils.getTokenSignedWithSecret(dataToTokenzie);

    it('returns the verified token with no errors', () => {
      const verify = require('./verify');
      const options = {
        keyStore: {
          getSecretOrPublicKey(keyId, callback) {
            return callback(null, 'super-secret-password');
          }
        }
      };
      verify.register(options, {}, () => {});
      verify.execute(token, (error, verifiedToken) => {
        expect(error).toBe(null);
        expect(verifiedToken).toMatchSnapshot();
      });
    });
  });

  describe('when given a token that has been tampered with', () => {
    const token = utils.getTokenThatHasBeenTamperedWith(dataToTokenzie);

    it('it returns an error with no token', () => {
      const verify = require('./verify');
      verify.register({ keys }, {}, () => {});
      verify.execute(token, (error, verifiedToken) => {
        expect(error).toMatchSnapshot();
        expect(verifiedToken).toBe(undefined);
      });
    });
  });

  describe('when given a string that is not a token', () => {
    const token = 'i am not a token';

    it('it returns an error with no token', () => {
      const verify = require('./verify');
      verify.register({ keys }, {}, () => {});
      verify.execute(token, (error, verifiedToken) => {
        expect(error).toMatchSnapshot();
        expect(verifiedToken).toBe(undefined);
      });
    });
  });

  describe('when given a token that is missing the key id', () => {
    const token = utils.getTokenWithMissingKeyId(dataToTokenzie);

    it('it returns an error with no token', () => {
      const verify = require('./verify');
      verify.register({ keys }, {}, () => {});
      verify.execute(token, (error, verifiedToken) => {
        expect(error).toMatchSnapshot();
        expect(verifiedToken).toBe(undefined);
      });
    });
  });
});
