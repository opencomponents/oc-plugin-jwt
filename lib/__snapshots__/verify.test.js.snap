// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`verify when a key is not found in the keystore it returns an error with no token 1`] = `[Error: Key no-key not found]`;

exports[`verify when a using a custom keystore returns the verified token with no errors 1`] = `
Object {
  "admin": true,
  "name": "John Doe",
  "sub": "1234567890",
}
`;

exports[`verify when given a string that is not a token it returns an error with no token 1`] = `[Error: Invalid Token]`;

exports[`verify when given a token that has been tampered with it returns an error with no token 1`] = `[JsonWebTokenError: invalid signature]`;

exports[`verify when given a token that is missing the key id it returns an error with no token 1`] = `[Error: Token missing key id (kid) in header]`;

exports[`verify when given a valid public key signed token returns the verified token with no errors 1`] = `
Object {
  "admin": true,
  "name": "John Doe",
  "sub": "1234567890",
}
`;

exports[`verify when given a valid secret signed token returns the verified token with no errors 1`] = `
Object {
  "admin": true,
  "name": "John Doe",
  "sub": "1234567890",
}
`;
